lib/rules/prefer-const.js

```diff
@@ -37,6 +41,51 @@ function getWriteReferenceIfOnce(variable) {
 }
 
 /**
+ * Checks whether or not a given reference is in a loop condition or a
+ * for-loop's updater.
+ *
+ * @param {escope.Reference} reference - A reference to check.
+ * @returns {boolean} `true` if the reference is in a loop condition or a
+ *      for-loop's updater.
+ */
+function isInLoopHead(reference) {
+    var node = reference.identifier;
+    var parent = node.parent;
+    var assignment = false;
+
+    while (parent) {
+        if (LOOP_TYPES.test(parent.type)) {
+            return true;
+        }
+
+        // VariableDeclaration can be at ForInStatement.left
+        // This is catching the code like `for (const {b = ++a} of foo()) { ... }`
+        if (assignment &&
+            parent.type === "VariableDeclaration" &&
+            FOR_IN_OF_TYPES.test(parent.parent.type) &&
+            parent.parent.left === parent
+        ) {
+            return true;
```

Comment:
This `a` should be caught:

```js
for (let a of foo()) { ... }
```

But this `a` should not be caught:

```js
let a;
for (let {b = a++} of foo()) { ... }
```

So the test of `ForInStatement.left` and `ForOfStatement.left` is complexity a bit.